README.md

```diff
@@ -4,6 +4,128 @@
+                   scsi::LunAddress lun)
+```
+
+This function takes in:
+- Raw SCSI command in bytes
+- Buffer which can either be an input buffer or an output buffer. As an input buffer, it will be where the data from the device is read into. As an output buffer, it stores variable-length command data or the data to be written to the device
+- Lun Address which is the NSID in NVMe
+
+From the parameters, the function then delegates to the appropriate command translator and then builds 1 or more NVMe commands. For some commands such as TestUnitReady or Report Supported Op Codes, the Translation Library does not build an NVMe command and instead writes data directly to the SCSI buffer.
+
+The functions returns:
+```
+BeginReponse {
+  ApiStatus status;   // Indicates correct usage of Translation Library 
+  uint32_t alloc_len; // Defines size of buffer passed to Translation::Complete
+}
+```
+### Translation::GetNvmeWrappers() ###
+```
+Span<NvmeCmdWrapper> GetNvmeWrappers()
+```
+This function returns the constructed NVMe commands along with other useful data.
+
+```
+NvmeCmdWrapper {
+  nvme::GenericQueueEntryCmd cmd; // NVMe command
+  uint32_t buffer_len; // length of buffer pointed to by the NVMe PRP pointer
+  bool is_admin; // whether command is an admin command. Needed for sending to correct NVMe queue
+}
+```
+
+### Translation::Complete() ###
+```
+CompleteReponse Complete(Span<const nvme::GenericQueueEntryCpl> cpl_data, 
+                         Span<const uint8_t> buffer_in,
+                         Span<const uint8_t> sense_buffer);
+```
+The function takes in:
+- NVMe Completion Queue Entry, which contains the NVMe response data
+- SCSI data in buffer where response data is written to
+- SCSI sense buffer where status codes are written to
+
+This function translates the NVMe response data and status codes and writes them to the appropriate buffer.
+
+### Translation::AbortPipeline ###
+```
+void AbortPipeline()
+```
+Finally, in the case that the Translation pipeline needs to be aborted, this function handles all the necesssary memory cleanup.
+
+### Intended Usage ###
+1. Get the Raw SCSI command and other data from the SCSI subsystem
+1. Pass data to Translation::Begin()
+1. Get constructed NVMe commands and other data with Translation::GetNvmeWrappers()
+1. Send the commands to the NVMe device for execution
+1. Pass the NVMe completion queue entries to Translation::Complete()
+
+## End-to-end translation ##
+
+### Setup ###
+1) Ensure the following:
+    - an NVMe device is attached to the machine at /dev/nvme0n1
+    - kernel version is >= 4.19
+1) Call `$ make` in the root directory. Ensure this results in a scsi2nvme.ko
+1) Call `$ insmod scsi2nvme.ko` to insert the kernel module.
+
+    Upon inserting, the SCSI mock device will be automatically attached. Once the device is attached, the SCSI subsystem will issue a set of commands. 
+
+    `$ lsscsi` should show the device with the name "NVMe &nbsp; &nbsp; &nbsp; &nbsp;"
```

Comment:
What's with all the `&nbsp;`?

Reply:
The name is supposed to be NVMe followed by 4 spaces and markdown will just render 4 consecutive spaces as one space. Although I realized that the way I had it gave it 8 spaces. I also added a comment in parenthesis to describe what it is

Reply:
Ah, nice!